grpc-for-smart-contracts.md

```diff
@@ -0,0 +1,263 @@
+- `address`: A null-terminated base58 encoding (same format as returned by `GetAccountList`) of an account address. 
+
+**Output:**
+
+- A JSON object containing information about the given address at the given block.
+  - Example:
+
+    ```json
+    {"accountAmount":"10000000000000","accountNonce":1,"accountEncryptedAmount":{"incomingAmounts":[],"selfAmount":"c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","startIndex":0},"accountEncryptionKey":"993fdc40bb8af4cb75caf8a53928d247be6285784b29578a06df312c28854c1bfac2fd0183967338b578772398d4120193c96e1a644ecb148615ec0940cdf30e2f373cc470f393231a4be681b8cbb27edf0b67fda7290a3545d62dc8038d1fa7","accountReleaseSchedule":{"schedule":[],"total":"0"},"accountCredentials":{"0":{"value":{"contents":{"ipIdentity":0,"regId":"93c96e1a644ecb148615ec0940cdf30e2f373cc470f393231a4be681b8cbb27edf0b67fda7290a3545d62dc8038d1fa7","policy":{"revealedAttributes":{},"createdAt":"202103","validTo":"202203"},"credentialPublicKeys":{"keys":{"0":{"verifyKey":"6bb6fb860514c7a1002fc9c3f3a9f7a7a518d54d0437b7dd7e9ba29c65263cd8","schemeId":"Ed25519"},"1":{"verifyKey":"80db411ba83f282ebe7ab3c037e481c72c460de1de87bc57a0b85a7dd347bd76","schemeId":"Ed25519"},"2":{"verifyKey":"c14a3e621d7ff5f1302a0b5c9371fbe8429e8db33b1ff33b962e468a901c159a","schemeId":"Ed25519"}},"threshold":2}},"type":"initial"},"v":0}}}
+    ```
+
+- Returns `null` if the address could not be found.
+
+## `GetInstances`
+
+*Retrieve the addresses of all smart contract instances.*
+
+```protobuf
+rpc GetInstances(BlockHash) returns (JsonResponse) {}
+
+message BlockHash {
+  string block_hash = 1;
+}
+```
+**Input:**
+
+- `block_hash`: A null-terminated base16 encoding of a block hash. 
+
+**Output:**
+
+- A JSON list of contract addresses on the chain.
+  - Example: `[{"subindex":0,"index":0},{"subindex":0,"index":1}]` 
+
+## `GetInstanceInfo`
+
+*Retrieve information about a given smart contract instance.*
+
+```protobuf
+rpc GetInstanceInfo(GetAddressInfoRequest) returns (JsonResponse) {}
+
+message GetAddressInfoRequest {
+  string block_hash = 1;
+  string address = 2;
+}
+```
+
+**Input:**
+
+- `block_hash`: A null-terminated base16 encoding of a block hash. 
+- `address`: A null-terminated JSON-encoded value (same format as returned by
+  `GetInstances`).
+
+**Output:**
+
+- A JSON object with information about the contract instance.
+  - Example: 
+
+  ```json
+  {"amount":"0","sourceModule":"609af98562314a5461990080df5514703bfb9f98d2c5b4535cf4b11f766cac29","owner":"4NyqjcJwKAckGE3gMGDJeqcyLZQxtfnDFTdvg74xYUU8myQTrB","methods":["counter.receive","counter.receive_optimized"],"name":"init_counter","model":"0000000000"}
```

Comment:
Same as above, please pretty print this, and it would be useful to comment about the fields, what they mean and what is always fixed, and what is just  given as an example.